### DISCUSSION TITLE
COMBO broadcasting via Anything Everywhere/?

### DISCUSSION THREAD

**Discussion by Gerkinfeltser** [chrisgoringe/cg-use-everywhere#178] [Ideas]
COMBO type outputs are so weird & kind of tedious to work with.
Being able to use COMBO output with UE nodes would be super great though if possible.
Right now they don't seem to play together it seems.

Here you can see the Anything Everywhere won't accept the COMBO. (The AE won't accept it):
![image](https://github.com/chrisgoringe/cg-use-everywhere/assets/1477787/406e7e2f-cfe8-499d-85f0-a04a8259d285)

**Comment by Gerkinfeltser** [comment]
Actually, this might just be an issue with primitive nodes in which case: Please Ignore! 😆

**Reply by chrisgoringe** [reply]
UE is very unlikely to ever work with combo or primitive inputs.

**Reply by Gerkinfeltser** [reply]
Understood! I assumed as much and found direct connection is possible so this can be disregarded. Thanks for the reply (& extension)!

**Comment by JorgeR81** [comment]
There is a workaround, with the `CR String To Combo` node from ComfyUI_Comfyroll_CustomNodes.
This should work for any COMBO ( e.g. AIO Aux Preprocessor ).

https://github.com/Fannovel16/comfyui_controlnet_aux/issues/289#issuecomment-2027461249

But we need to use the Simple String node for the UE? input, so there will be no dropdown list.

**Comment by JorgeR81** [comment]
Could it be possible to get the COMBO list connected to the `CR String To Combo` node and put that on an autocomplete widget, on the Simple String node ?

If so, it would also be worth it to recreate the String To Combo node in the UE? suite.

**Comment by chrisgoringe** [comment]
I'm working on this - inspired by this workaround I think I've found a way. 

At present, this works (in a dev branch): convert combo to string, UE it as a string, and convert back (I've replicated an ultra-simple version of String to Combo).

![Screenshot 2024-04-03 115241](https://github.com/chrisgoringe/cg-use-everywhere/assets/4371971/5aed73c9-ad8d-4cd5-95a7-02d52d4193c5)

But the bypassed kSample is still required. Working on that at the moment.

**Comment by chrisgoringe** [comment]
OK - anyone want to try this out on a branch?

There's a branch called "combos" which seems to be working - like the image above, but you don't need the bypassed node to hang around any more.

**Comment by JorgeR81** [comment]
It's working for me !

But I need to connect the nodes in a specific sequence, for this to work. 

If I connect the "basic" Primitive to the `Combo to String` node, the Primitive will not "convert" and the `Combo to String` node will get "glued" to the mouse cursor.
( I need to click the browser's refresh button, to set the `Combo to String` node free ).

![test5](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/674c9911-5674-4e9f-afe0-07e6c76273b7)


For this to work, need to connect the "basic" Primitive to the KSampler, for it to "convert" to the right kind of Primitive.

![test2](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/db38ff03-af77-4eca-989f-eb44b565b5bb)

Then I can also connect the new Primitive to `Combo to String` ...

![test3](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/b08eca43-8bea-4fc5-9a8a-2ddf2f6a9668)

... and finally, connect `String to Combo` to the KSampler.

![test4](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/a7b2de66-0cf0-464f-be74-7be3ce9810f4)

**Comment by chrisgoringe** [comment]
Yes, that sequence is required for the primitive to 'learn' what sort of COMBO to be. The glueing is a bug, though - it ought to refuse to connect to the Combo to String node before it's been 'taught'.

You know you can create the right sort of primitive by double-clicking on the unconnected input?

**Comment by JorgeR81** [comment]
So the most efficient way to do this would be to create an UE + combos template ( like on the image ), that we can load when necessary. 

Double click the unconnected input, to create the Primitive, connected to the "target" input.

And make the 2 required connections:

- Primitive to `Combo to String`.

- And then `String to Combo` to the "target" input, to replace the original Primitive connection. 



Works for me !

![template](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/f4130b08-576c-4ecf-9d68-079f92d46c90)

**Comment by JorgeR81** [comment]
It also works for multiple inputs. 

We just need to rename the `String to Combo` inputs. 
This way the autocomplete also works. 

![test03](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/664f2750-3aed-4d78-95b1-905430dfbab0)

![test02](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/1558b4b0-d841-4782-9351-aa5c344aecd0)

**Comment by chrisgoringe** [comment]
Just pushed a small update to fix the glue problem.

Doesn't work if the primitive and combo to string nodes are both part of a template though

**Comment by JorgeR81** [comment]
> fix the glue problem

Yes, that's also fixed for me.

But I also noticed that if the `String to Combo` node is created from de node suggestions, the connection will not be completely successful. 

There is an * in the UE? node input.
And the "taught" Primitive will not connect to the  `String to Combo` node.

![test001](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/98e8b693-1d8e-42ac-9505-6c11e5c44f87)

![test002](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/7860355f-b33f-4cca-b46d-ab2ee552434c)

**Comment by JorgeR81** [comment]
> Doesn't work if the primitive and combo to string nodes are both part of a template though

It's OK, about the template I meant the UE? node + the 2 new combo nodes ( the ones on the group, in the image ).

This still works for me.

![test003](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/3b045be1-0128-4341-811a-ed46b3ce7af3)

**Comment by JorgeR81** [comment]
There may be an even better solution for COMBO broadcasting !?

I'm having some issues connecting to the Trung0246 nodes ( I've made an issue for that: https://github.com/chrisgoringe/cg-use-everywhere/issues/181 )

But I was able to broadcast a batch of COMBO values, in a XY plot workflow, without the new combo branch ! 

The connection was named `*` in the UE? node instead of COMBO, when I connected the Merge node form Trung0246

But I can only broadcast to the Highway Bach node, not to the KSampler directly. 

![hb](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/75c59d79-10bf-4a65-b3d4-969b3d6c6b10)

In a comment about another issue I had, @Trung0246 said this wildcard `*` works for anything including COMBO.

https://github.com/Trung0246/ComfyUI-0246/issues/15#issuecomment-2046310539

In some cases this does not work ( COMBO is needed, and `*` gives an error ), but in this case ( with UE? nodes ), it seems to work. 
The `*` seems to allow UE? nodes to broadcast COMBO data, without the nodes of the combo branch.

**Comment by chrisgoringe** [comment]
The issue the combo branch nodes are dealing with is not so much COMBO, as it is *primitives* - the nodes auto-created when you turn a combo widget into an input and double-click it are COMBO primitives. And the issue with primitives is that they don't remember what they are... so when you reload, if they aren't connected to something, they lose their type.

So the combo branch is all about UE nodes being able to remember what they were getting from the primitive, so that when reloaded, the primitive can know what it's supposed to send!

**Comment by JorgeR81** [comment]
Yes, of course, you're right !
I already knew this, ... but got confused for a minute. 

The UE? nodes are already able to broadcast COMBO.

The problem is that there are no dedicated COMBO selectors for all possible cases, like this one from rgthree.
So we need to use Primitives ...

![rg3](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/b873eac8-24da-453d-9a22-6c1f50c30fa3)

**Comment by JorgeR81** [comment]
Some of the Trung0246 nodes remind me of Primitive nodes, in the sense that they can have different output types, according to the node that they are connected to.
Like this Junction Batch node.
Although it's not the same thing ...


![junc](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/d04b97b4-0de8-4e11-ba59-cc39043eaaa4)

But currently we also need a trick to create the right type of output, to make this work.
First we need to connect to a node with a COMBO input, like we would do with a Primitive.
Only then we can make the connection to the Merge node, for the input to also be COMBO.

![cp](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/5b1b2e9d-324a-46ec-b66f-f839f132e283)

**Comment by JorgeR81** [comment]
Hey, how is this new feature doing ?

Did you find some major hurdles in the development, 

or you just haven't had the time to work on it ?
